File: scripts/license_check.py
"""CI step 1 — dependency licence gate (PHASE-0 Task 1).

Walks the *resolved* dependency tree via importlib.metadata (not the declared
one — transitive pulls are exactly what this is meant to catch), classifies each
distribution against scripts/license_policy.toml, and exits non-zero on anything
unknown or denied.

Usage:
    python scripts/license_check.py [--root tracebed] [--policy scripts/license_policy.toml]
    python scripts/license_check.py --self-test    # proves the gate actually fails
"""

from __future__ import annotations

import argparse
import re
import sys
import tomllib
from dataclasses import dataclass, field
from importlib import metadata
from pathlib import Path

REPO_ROOT = Path(__file__).resolve().parent.parent

# A gate that crashes while PRINTING its own (passing) result is a false red.
# Windows consoles default to cp1252; these reports use box drawing and em
# dashes, and `print` raised UnicodeEncodeError AFTER the gate had already
# decided its verdict -- so a PASS run looked like a gate failure. Force
# UTF-8 on the streams rather than downgrading the output, and fall back to
# replacement characters if even that is refused: the verdict must survive the
# terminal it is printed to.
for _stream in (sys.stdout, sys.stderr):
    if hasattr(_stream, "reconfigure"):
        _stream.reconfigure(encoding="utf-8", errors="replace")


# Distributions that are part of the toolchain rather than the shipped artefact.
# They are still reported, but a denial here is a warning, not a gate failure.
DEV_ONLY = {
    "pytest",
    "pytest-asyncio",
    "mypy",
    "mypy-extensions",
    "ruff",
    "iniconfig",
    "pluggy",
    "coverage",
    "hatchling",
    "pathspec",
    "trove-classifiers",
    "editables",
}


@dataclass
class Policy:
    allow: set[str]
    conditional: set[str]
    deny: set[str]
    lgpl_rationale: dict[str, str]
    manual_overrides: dict[str, str]

    @staticmethod
    def load(path: Path) -> Policy:
        raw = tomllib.loads(path.read_text(encoding="utf-8"))
        return Policy(
            allow={_norm_licence(x) for x in raw.get("allow", [])},
            conditional={_norm_licence(x) for x in raw.get("conditional", [])},
            deny={_norm_licence(x) for x in raw.get("deny", [])},
            lgpl_rationale={_norm_dist(k): v for k, v in raw.get("lgpl_rationale", {}).items()},
            manual_overrides={_norm_dist(k): v for k, v in raw.get("manual_overrides", {}).items()},
        )


@dataclass
class Finding:
    dist: str
    version: str
    licence: str
    status: str  # allowed | conditional-ok | denied | unknown
    detail: str = ""


@dataclass
class Report:
    findings: list[Finding] = field(default_factory=list)

    @property
    def failures(self) -> list[Finding]:
        return [
            f
            for f in self.findings
            if f.status in {"denied", "unknown"} and _norm_dist(f.dist) not in DEV_ONLY
        ]


def _norm_licence(value: str) -> str:
    # Parentheses are dropped rather than stripped from the edges: classifier strings
    # like "Mozilla Public License 2.0 (MPL 2.0)" carry them mid-string, where an
    # edge-strip leaves an unbalanced atom that can never match the policy.
    return re.sub(r"[\s_]+", "-", re.sub(r"[()]", " ", value).strip().lower())


def _norm_dist(value: str) -> str:
    return re.sub(r"[-_.]+", "-", value.strip().lower())


_CLASSIFIER_LICENCE = re.compile(r"^License\s*::\s*(?:OSI Approved\s*::\s*)?(.+)$")


def read_licence(dist: metadata.Distribution) -> str:
    """Best-effort licence extraction across the three metadata conventions."""
    meta = dist.metadata

    # PEP 639: License-Expression is authoritative when present.
    expr = meta.get("License-Expression")
    if expr:
        return str(expr).strip()

    classifiers = [str(c) for c in meta.get_all("Classifier") or []]
    from_classifiers = [
        m.group(1).strip() for c in classifiers if (m := _CLASSIFIER_LICENCE.match(c))
    ]
    # "OSI Approved" alone carries no information; drop it.
    from_classifiers = [c for c in from_classifiers if c.lower() != "osi approved"]
    if from_classifiers:
        return " OR ".join(from_classifiers)

    legacy = meta.get("License")
    if legacy:
        text = str(legacy).strip()
        # Some projects dump the whole licence text into this field.
        return text if len(text) < 120 else text.splitlines()[0][:120]

    return ""


def split_expression(expr: str) -> list[str]:
    """Split an SPDX-ish expression into its atoms. `A OR B` passes if any atom passes."""
    parts = re.split(r"\s+(?:OR|AND)\s+|\s*;\s*", expr, flags=re.IGNORECASE)
    return [p.strip(" ()") for p in parts if p.strip(" ()")]


def classify(dist_name: str, licence: str, policy: Policy) -> Finding:
    key = _norm_dist(dist_name)
    version = _version_of(dist_name)

    if key in policy.manual_overrides:
        return Finding(dist_name, version, licence or "(none)", "allowed",
                       f"manual override: {policy.manual_overrides[key]}")

    if not licence:
        return Finding(dist_name, version, "(none)", "unknown",
                       "no License-Expression, no License:: classifier, no License field")

    atoms = [_norm_licence(a) for a in split_expression(licence)]

    if any(a in policy.deny for a in atoms):
        denied = next(a for a in atoms if a in policy.deny)
        return Finding(dist_name, version, licence, "denied", f"denylisted licence: {denied}")

    if any(a in policy.allow for a in atoms):
        return Finding(dist_name, version, licence, "allowed")

    if any(a in policy.conditional for a in atoms):
        if key in policy.lgpl_rationale:
            return Finding(dist_name, version, licence, "conditional-ok",
                           policy.lgpl_rationale[key])
        return Finding(dist_name, version, licence, "denied",
                       "conditional licence with no entry in [lgpl_rationale]")

    return Finding(dist_name, version, licence, "unknown",
                   "licence not present in allow, conditional, or deny lists")


def _version_of(dist_name: str) -> str:
    try:
        return metadata.version(dist_name)
    except metadata.PackageNotFoundError:
        return "?"


def resolve_tree(root: str) -> set[str]:
    """Transitive closure of Requires-Dist starting at `root`, extras included."""
    seen: set[str] = set()
    queue = [root]
    while queue:
        name = queue.pop()
        key = _norm_dist(name)
        if key in seen:
            continue
        try:
            dist = metadata.distribution(name)
        except metadata.PackageNotFoundError:
            continue
        seen.add(key)
        for req in dist.requires or []:
            dep = re.split(r"[\s\[;(<>=!~]", req.strip(), maxsplit=1)[0]
            if dep:
                queue.append(dep)
    return seen


def collect(root: str | None, policy: Policy) -> Report:
    if root:
        wanted = resolve_tree(root)
        if not wanted:
            print(f"warning: '{root}' is not installed; falling back to the full environment",
                  file=sys.stderr)
    else:
        wanted = set()

    report = Report()
    for dist in metadata.distributions():
        name = dist.metadata.get("Name")
        if not name:
            continue
        if wanted and _norm_dist(name) not in wanted:
            continue
        report.findings.append(classify(str(name), read_licence(dist), policy))
    report.findings.sort(key=lambda f: (f.status != "denied", f.status != "unknown", f.dist.lower()))
    return report


def render(report: Report) -> str:
    lines = ["dist                           version      status          licence"]
    lines.append("-" * 96)
    for f in report.findings:
        lines.append(f"{f.dist[:30]:<30} {f.version[:12]:<12} {f.status:<15} {f.licence[:40]}")
        if f.detail and f.status != "allowed":
            lines.append(f"{'':<30} {'':<12} └─ {f.detail}")
    return "\n".join(lines)


def self_test(policy: Policy) -> int:
    """Prove the gate bites: a fake SSPL distribution must be rejected."""
    cases = [
        ("fake-sspl-db", "SSPL-1.0", "denied"),
        ("fake-agpl-cache", "AGPL-3.0-only", "denied"),
        ("psycopg", "LGPL-3.0-only", "conditional-ok"),
        ("some-other-lgpl", "LGPL-3.0-only", "denied"),
        ("fastapi", "MIT", "allowed"),
        ("mystery-lib", "", "unknown"),
        ("dual-licensed", "Apache-2.0 OR MIT", "allowed"),
    ]
    ok = True
    for name, licence, expected in cases:
        got = classify(name, licence, policy).status
        mark = "ok " if got == expected else "FAIL"
        if got != expected:
            ok = False
        print(f"[{mark}] {name:<20} {licence or '(none)':<22} expected={expected:<15} got={got}")
    return 0 if ok else 1


# Names that appear in `pyproject.toml` but are exempt from the DECISIONS.md audit below,
# with the reason. Deliberately tiny: the point of the audit is that an addition is a
# reviewable event, and an exemption list that grows is the audit failing quietly.
_DEPENDENCY_AUDIT_EXEMPT: dict[str, str] = {
    # Build backend, not a runtime or test dependency of the service.
    "hatchling": "build backend, declared in [build-system]",
}


def _declared_dependencies() -> set[str]:
    """Every distribution name in `pyproject.toml`'s runtime and dev dependency lists."""
    import re
    import tomllib

    raw = (REPO_ROOT / "pyproject.toml").read_bytes()
    data = tomllib.loads(raw.decode("utf-8"))
    project = data.get("project", {})
    specs: list[str] = list(project.get("dependencies", []))
    for extra in project.get("optional-dependencies", {}).values():
        specs.extend(extra)
    names: set[str] = set()
    for spec in specs:
        # "psycopg[binary,pool]>=3.2" -> "psycopg"
        name = re.split(r"[\[<>=!~;\s]", spec, maxsplit=1)[0].strip()
        if name:
            names.add(_norm_dist(name))
    return names


def dependency_audit() -> int:
    """Hard rule 10 / D-036: every declared dependency carries a DECISIONS.md entry.

    Three dependencies (`prometheus-client`, `pytest-asyncio`, `types-pyyaml`) reached this
    tree with no entry, while D-036's own inventory named a package (`onnxruntime`) that was
    never installed — the inventory was wrong in both directions at once. A grep is a weak
    check and it is the right strength here: the question is "did a human write this down",
    not "is the prose correct".
    """
    decisions = (REPO_ROOT / "DECISIONS.md").read_text(encoding="utf-8").lower()
    missing = sorted(
        name
        for name in _declared_dependencies()
        if name not in _DEPENDENCY_AUDIT_EXEMPT
        and name not in decisions
        and name.replace("-", "_") not in decisions
    )
    for name in missing:
        print(f"undeclared dependency (no DECISIONS.md entry): {name}")
    print()
    if missing:
        print(f"DEPENDENCY AUDIT: FAIL — {len(missing)} dependency(ies) with no decision record")
        return 1
    print("DEPENDENCY AUDIT: PASS — every declared dependency is named in DECISIONS.md")
    return 0


def main(argv: list[str] | None = None) -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument(
        "--dependency-audit",
        action="store_true",
        help="check that every pyproject dependency is named in DECISIONS.md (hard rule 10)",
    )
    ap.add_argument("--root", default="tracebed",
                    help="walk only this distribution's dependency closure (default: tracebed)")
    ap.add_argument("--all", action="store_true", help="check every installed distribution")
    ap.add_argument("--policy", type=Path, default=REPO_ROOT / "scripts" / "license_policy.toml")
    ap.add_argument("--self-test", action="store_true")
    args = ap.parse_args(argv)

    if args.dependency_audit:
        return dependency_audit()

    policy = Policy.load(args.policy)
    if args.self_test:
        return self_test(policy)

    report = collect(None if args.all else args.root, policy)
    print(render(report))

    failures = report.failures
    print()
    if failures:
        print(f"LICENCE GATE: FAIL — {len(failures)} distribution(s) blocked")
        for f in failures:
            print(f"  - {f.dist} {f.version}: {f.licence or '(none)'} — {f.detail}")
        return 1
    print(f"LICENCE GATE: PASS — {len(report.findings)} distribution(s) cleared")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: dashboard/scripts/license_check.mjs
#!/usr/bin/env node
/**
 * Dashboard dependency licence gate — the npm counterpart of
 * scripts/license_check.py, and deliberately the same discipline:
 *
 *   - it walks the RESOLVED tree on disk (node_modules), not the declared one
 *     in package.json, because transitive pulls are exactly what this catches;
 *   - an UNKNOWN licence fails. Silence is not consent. A package that ships no
 *     machine-readable licence is treated as more dangerous than one that
 *     honestly declares a copyleft licence, because nobody has read it;
 *   - a conditional licence passes only when THIS distribution is named in
 *     RATIONALE with a written justification (the psycopg carve-out pattern
 *     from scripts/license_policy.toml, D-036);
 *   - `--self-test` proves the gate bites, so a gate that has silently stopped
 *     classifying anything cannot masquerade as a clean run.
 *
 * The policy lives in this file rather than beside license_policy.toml because
 * Node has no TOML parser in its standard library, and a licence gate that
 * needs a dependency in order to check dependencies is a gate with a hole in
 * it. The two policies are kept deliberately consistent; where they differ
 * (CC0-1.0, CC-BY-4.0) it is because those licences appear only in the npm
 * tree and never in the Python one.
 *
 * Usage:
 *   node scripts/license_check.mjs              # gate the dashboard's tree
 *   node scripts/license_check.mjs --self-test  # prove the gate actually fails
 *   node scripts/license_check.mjs --all        # include dev-only, same verdict
 */

import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { dirname, join, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const HERE = dirname(fileURLToPath(import.meta.url));
const DASHBOARD_ROOT = resolve(HERE, "..");

// --------------------------------------------------------------------- //
// Policy
// --------------------------------------------------------------------- //

/** Accepted outright. Permissive: no obligation beyond preserving a notice. */
const ALLOW = new Set(
  [
    "MIT",
    "MIT-0",
    "ISC",
    "BSD-2-Clause",
    "BSD-3-Clause",
    "0BSD",
    "BSD Zero Clause License",
    "Apache-2.0",
    "Unlicense",
    "Python-2.0",
    "PSF-2.0",
    // File-level copyleft: obliges publishing modifications to the covered
    // FILES, not relicensing the dashboard. Nothing here is vendored or
    // modified, so the obligation never attaches. Same reasoning, same
    // wording, as the Python policy's MPL entry.
    "MPL-2.0",
  ].map(normaliseLicence)
);

/** Accepted only when the package is named in RATIONALE below. */
const CONDITIONAL = new Set(
  [
    // Public-domain dedication. Strictly MORE permissive than MIT — it is
    // conditional here not because it is risky but because it is not on the
    // mandated allowlist, and widening that list silently is how an allowlist
    // stops meaning anything. Each use gets a named, recorded assertion.
    "CC0-1.0",
    // Attribution required on redistribution of the WORK. Fine for a build
    // tool consulted at compile time, not fine for anything vendored into a
    // shipped bundle — which is precisely the distinction RATIONALE records.
    "CC-BY-4.0",
    "LGPL-3.0-only",
    "LGPL-3.0",
    "LGPL-2.1",
  ].map(normaliseLicence)
);

/** Hard fail. No rationale can rescue these — network copyleft, source-available
 * licences, and the "no licence at all" marker npm uses. */
const DENY = new Set(
  [
    "AGPL-1.0",
    "AGPL-3.0",
    "AGPL-3.0-only",
    "AGPL-3.0-or-later",
    "GPL-2.0",
    "GPL-2.0-only",
    "GPL-3.0",
    "GPL-3.0-only",
    "GPL-3.0-or-later",
    "SSPL",
    "SSPL-1.0",
    "BUSL-1.1",
    "BSL-1.1",
    "Elastic-2.0",
    "RSAL",
    "CC-BY-NC-4.0",
    "CC-BY-SA-4.0",
    "UNLICENSED",
  ].map(normaliseLicence)
);

/**
 * package name -> why its conditional licence is acceptable HERE.
 *
 * Every entry is a human assertion. It must say what the package is, why the
 * licence's obligation does not attach, and — critically — whether the package
 * reaches the shipped bundle, because that is the fact the obligation turns on.
 */
const RATIONALE = {
  "caniuse-lite":
    "CC-BY-4.0. Browser-support DATA consulted by autoprefixer/browserslist at " +
    "build time to choose CSS prefixes. Build-time toolchain only: the database " +
    "is never copied into dist/, so the dashboard redistributes none of the " +
    "CC-BY-licensed work and the attribution obligation does not attach to the " +
    "shipped artefact. Attribution is preserved in node_modules for anyone " +
    "building from source.",
  "language-subtag-registry":
    "CC0-1.0. IANA subtag data used by eslint-plugin-jsx-a11y to validate lang " +
    "attributes. Lint-time only, never bundled. CC0 is a public-domain " +
    "dedication carrying no obligation even if it were redistributed.",
};

/** Packages reachable only through devDependencies are toolchain, not artefact.
 * They are still classified and still reported; this flag only decides whether
 * a failure among them blocks the gate, mirroring the Python gate's DEV_ONLY
 * set. Default is STRICT (they block too) — the lenient path must be asked for. */
const DEV_FAILURES_BLOCK = true;

// --------------------------------------------------------------------- //
// Licence parsing
// --------------------------------------------------------------------- //

function normaliseLicence(value) {
  // Parentheses are dropped mid-string rather than trimmed from the edges, for
  // the same reason the Python gate does it: "(MIT OR CC0-1.0)" must not leave
  // an unbalanced atom that can never match anything.
  return String(value)
    .replace(/[()]/g, " ")
    .trim()
    .toLowerCase()
    .replace(/[\s_]+/g, "-");
}

/** Read a licence out of the three package.json conventions, newest first. */
function readLicence(pkg) {
  if (typeof pkg.license === "string") return pkg.license.trim();
  // Legacy object form: { "license": { "type": "MIT", "url": ... } }
  if (pkg.license && typeof pkg.license === "object" && typeof pkg.license.type === "string") {
    return pkg.license.type.trim();
  }
  // Legacy array form: { "licenses": [{ "type": "MIT" }, ...] } — an OR by
  // npm's own historical definition, so join it as one.
  if (Array.isArray(pkg.licenses)) {
    const types = pkg.licenses
      .map((l) => (typeof l === "string" ? l : l && typeof l.type === "string" ? l.type : ""))
      .filter(Boolean);
    if (types.length > 0) return types.join(" OR ");
  }
  return "";
}

/** Split an SPDX expression into atoms. `A OR B` passes if ANY atom passes;
 * `A AND B` is treated the same way here, conservatively flagged by the DENY
 * check running first — a denied atom anywhere fails regardless of operator. */
function splitExpression(expr) {
  return expr
    .split(/\s+(?:OR|AND)\s+|\s*;\s*/i)
    .map((p) => p.replace(/^[\s()]+|[\s()]+$/g, ""))
    .filter(Boolean);
}

/** A licence string that names a file instead of a licence tells us nothing
 * machine-readable, and "nothing" must not be mistaken for a match. */
function isUninformative(licence) {
  return /^see\s+license/i.test(licence) || licence === "" || /^custom$/i.test(licence);
}

function classify(name, licence, channel) {
  if (isUninformative(licence)) {
    return {
      name,
      licence: licence || "(none)",
      channel,
      status: "unknown",
      detail: "no machine-readable licence in package.json (no license, no licenses[])",
    };
  }

  const atoms = splitExpression(licence).map(normaliseLicence);

  const denied = atoms.find((a) => DENY.has(a));
  if (denied) {
    return { name, licence, channel, status: "denied", detail: `denylisted licence: ${denied}` };
  }
  if (atoms.some((a) => ALLOW.has(a))) {
    return { name, licence, channel, status: "allowed", detail: "" };
  }
  if (atoms.some((a) => CONDITIONAL.has(a))) {
    const why = RATIONALE[name];
    if (why) return { name, licence, channel, status: "conditional-ok", detail: why };
    return {
      name,
      licence,
      channel,
      status: "denied",
      detail: "conditional licence with no named entry in RATIONALE",
    };
  }
  return {
    name,
    licence,
    channel,
    status: "unknown",
    detail: "licence not present in ALLOW, CONDITIONAL, or DENY",
  };
}

// --------------------------------------------------------------------- //
// Resolved-tree walk
// --------------------------------------------------------------------- //

function readJson(path) {
  return JSON.parse(readFileSync(path, "utf8"));
}

/** Node's own resolution: walk up from `fromDir` looking for node_modules/<name>.
 * This is what actually gets loaded, which is the only tree worth gating —
 * a hoisted transitive dep and a top-level one are the same risk. */
function resolvePackageDir(name, fromDir) {
  let dir = fromDir;
  for (;;) {
    const candidate = join(dir, "node_modules", name);
    if (existsSync(join(candidate, "package.json"))) return candidate;
    const parent = dirname(dir);
    if (parent === dir) return null;
    dir = parent;
  }
}

/**
 * BFS from the root manifest. `runtime` marks everything reachable through
 * `dependencies` alone — the closure that can actually reach dist/. Anything
 * reached only via devDependencies is toolchain. Runtime always wins a tie,
 * because a package that is both must be judged by its stricter role.
 */
function walkTree(rootDir) {
  const manifest = readJson(join(rootDir, "package.json"));
  const found = new Map(); // name -> { version, licence, channel }
  const queue = [];

  for (const name of Object.keys(manifest.dependencies ?? {})) {
    queue.push({ name, from: rootDir, channel: "runtime" });
  }
  for (const name of Object.keys(manifest.devDependencies ?? {})) {
    queue.push({ name, from: rootDir, channel: "dev" });
  }

  while (queue.length > 0) {
    const { name, from, channel } = queue.shift();
    const seen = found.get(name);
    // Re-walk only when a dev-marked package turns out to be runtime-reachable.
    if (seen && !(seen.channel === "dev" && channel === "runtime")) continue;

    const dir = resolvePackageDir(name, from);
    if (!dir) {
      // A declared dependency that is not on disk means the tree was never
      // installed, or was installed partially. Either way the gate has not
      // seen what will actually ship, and saying PASS would be a lie.
      found.set(name, { version: "(not installed)", licence: "", channel, missing: true });
      continue;
    }
    const pkg = readJson(join(dir, "package.json"));
    found.set(name, {
      version: pkg.version ?? "?",
      licence: readLicence(pkg),
      channel,
      dir,
    });

    for (const dep of Object.keys(pkg.dependencies ?? {})) {
      queue.push({ name: dep, from: dir, channel });
    }
    // Optional deps that are present on disk do load, so they are gated too.
    for (const dep of Object.keys(pkg.optionalDependencies ?? {})) {
      if (resolvePackageDir(dep, dir)) queue.push({ name: dep, from: dir, channel });
    }
  }
  return found;
}

/** Everything physically installed, whether or not the manifest closure reaches
 * it. Used to report drift between "what we declared" and "what npm put on
 * disk" — a package present but unreferenced is still a package a build can
 * pick up. */
function scanInstalled(rootDir) {
  const modules = join(rootDir, "node_modules");
  const out = new Set();
  if (!existsSync(modules)) return out;
  const visit = (dir, scope) => {
    for (const entry of readdirSync(dir)) {
      if (entry.startsWith(".")) continue;
      const p = join(dir, entry);
      if (!statSync(p).isDirectory()) continue;
      if (entry.startsWith("@")) {
        visit(p, entry);
        continue;
      }
      if (existsSync(join(p, "package.json"))) out.add(scope ? `${scope}/${entry}` : entry);
    }
  };
  visit(modules, null);
  return out;
}

// --------------------------------------------------------------------- //
// Report
// --------------------------------------------------------------------- //

const STATUS_ORDER = { denied: 0, unknown: 1, "conditional-ok": 2, allowed: 3 };

function render(findings) {
  const lines = [
    "package                                  version      channel   status           licence",
    "-".repeat(110),
  ];
  for (const f of findings) {
    lines.push(
      `${f.name.slice(0, 40).padEnd(40)} ${String(f.version).slice(0, 12).padEnd(12)} ` +
        `${f.channel.padEnd(9)} ${f.status.padEnd(16)} ${f.licence.slice(0, 30)}`
    );
    if (f.detail && f.status !== "allowed") {
      const wrapped = f.detail.length > 88 ? `${f.detail.slice(0, 88)}...` : f.detail;
      lines.push(`${"".padEnd(40)} ${"".padEnd(12)} '-- ${wrapped}`);
    }
  }
  return lines.join("\n");
}

// --------------------------------------------------------------------- //
// Self-test
// --------------------------------------------------------------------- //

/**
 * A gate nobody has watched fail is a gate nobody knows is running. Each case
 * below is a licence class the dashboard could plausibly acquire, and the two
 * that matter most are the last two: a package with NO licence must fail, and
 * a conditional licence on an UNNAMED package must fail. Those are the paths
 * where a broken gate quietly returns "allowed".
 */
function selfTest() {
  const cases = [
    ["fake-sspl-db", "SSPL-1.0", "runtime", "denied"],
    ["fake-agpl-widget", "AGPL-3.0-only", "runtime", "denied"],
    ["fake-gpl-lib", "GPL-3.0", "runtime", "denied"],
    ["fake-busl-lib", "BUSL-1.1", "runtime", "denied"],
    ["react", "MIT", "runtime", "allowed"],
    ["some-apache-lib", "Apache-2.0", "runtime", "allowed"],
    ["type-fest", "(MIT OR CC0-1.0)", "dev", "allowed"],
    ["caniuse-lite", "CC-BY-4.0", "dev", "conditional-ok"],
    ["language-subtag-registry", "CC0-1.0", "dev", "conditional-ok"],
    ["unnamed-cc0-lib", "CC0-1.0", "dev", "denied"],
    ["unnamed-lgpl-lib", "LGPL-3.0-only", "runtime", "denied"],
    ["mystery-lib", "", "runtime", "unknown"],
    ["file-licence-lib", "SEE LICENSE IN LICENSE.md", "runtime", "unknown"],
    ["private-lib", "UNLICENSED", "runtime", "denied"],
    ["exotic-lib", "WTFPL", "runtime", "unknown"],
    // A denied atom must lose even when OR-ed with a permissive one: the
    // permissive branch does not erase the fact that we cannot tell which
    // branch a downstream consumer will rely on.
    ["dual-trap", "MIT OR SSPL-1.0", "runtime", "denied"],
  ];

  let ok = true;
  for (const [name, licence, channel, expected] of cases) {
    const got = classify(name, licence, channel).status;
    if (got !== expected) ok = false;
    console.log(
      `[${got === expected ? "ok  " : "FAIL"}] ${name.padEnd(26)} ` +
        `${(licence || "(none)").padEnd(26)} expected=${expected.padEnd(15)} got=${got}`
    );
  }
  console.log();
  console.log(ok ? "SELF-TEST: PASS — the gate rejects what it claims to reject" : "SELF-TEST: FAIL");
  return ok ? 0 : 1;
}

// --------------------------------------------------------------------- //
// Main
// --------------------------------------------------------------------- //

function main(argv) {
  if (argv.includes("--self-test")) return selfTest();

  if (!existsSync(join(DASHBOARD_ROOT, "node_modules"))) {
    console.error(
      "LICENCE GATE: FAIL — dashboard/node_modules does not exist. Run `npm ci` first;\n" +
        "a gate that has nothing to inspect must not report PASS."
    );
    return 1;
  }

  const tree = walkTree(DASHBOARD_ROOT);
  const findings = [];
  for (const [name, info] of tree) {
    if (info.missing) {
      findings.push({
        name,
        version: "(not installed)",
        licence: "(none)",
        channel: info.channel,
        status: "unknown",
        detail: "declared as a dependency but not present in node_modules",
      });
      continue;
    }
    findings.push({ ...classify(name, info.licence, info.channel), version: info.version });
  }
  findings.sort(
    (a, b) => STATUS_ORDER[a.status] - STATUS_ORDER[b.status] || a.name.localeCompare(b.name)
  );

  console.log(render(findings));

  const installed = scanInstalled(DASHBOARD_ROOT);
  const unreached = [...installed].filter((n) => !tree.has(n));

  const blocking = findings.filter(
    (f) =>
      (f.status === "denied" || f.status === "unknown") &&
      (DEV_FAILURES_BLOCK || f.channel === "runtime")
  );

  const counts = findings.reduce((acc, f) => {
    acc[f.status] = (acc[f.status] ?? 0) + 1;
    return acc;
  }, {});

  console.log();
  console.log(
    `${findings.length} package(s) in the resolved tree: ` +
      `${counts.allowed ?? 0} allowed, ${counts["conditional-ok"] ?? 0} conditional (named), ` +
      `${counts.denied ?? 0} denied, ${counts.unknown ?? 0} unknown.`
  );
  if (unreached.length > 0) {
    // Not a failure: npm hoists build-time-only packages that no manifest in
    // the closure references. Reported because an unreferenced package on disk
    // is still something a future import can reach without review.
    console.log(
      `${unreached.length} installed package(s) are not reachable from package.json's ` +
        "closure (npm hoisting artefacts); they ship nothing and are not gated."
    );
  }

  if (blocking.length > 0) {
    console.log(`\nLICENCE GATE: FAIL — ${blocking.length} package(s) blocked`);
    for (const f of blocking) {
      console.log(`  - ${f.name} ${f.version} [${f.channel}]: ${f.licence} — ${f.detail}`);
    }
    return 1;
  }
  console.log(`\nLICENCE GATE: PASS — ${findings.length} package(s) cleared`);
  return 0;
}

process.exit(main(process.argv.slice(2)));
